src/state.rs

```diff
@@ -0,0 +1,86 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use dusk_plonk::prelude::*;
+use dusk_poseidon::sponge;
+use dusk_poseidon::tree::{PoseidonBranch, PoseidonLeaf, PoseidonTree};
+use nstack::annotation::Keyed;
+
+use crate::license::License;
+
+pub struct State<const DEPTH: usize> {
+    pub tree: PoseidonTree<DataLeaf, (), DEPTH>,
+}
+
+impl<const DEPTH: usize> Default for State<DEPTH> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<const DEPTH: usize> State<DEPTH> {
+    pub fn new() -> State<DEPTH> {
+        State {
+            tree: PoseidonTree::<DataLeaf, (), DEPTH>::default(),
+        }
+    }
+    pub fn append_license(&mut self, lic: &License) {
+        let lpk = JubJubAffine::from(*lic.lsa.pk_r().as_ref());
+        let license_hash = sponge::hash(&[lpk.get_x(), lpk.get_y()]);
+
+        self.tree.push(DataLeaf::new(license_hash));
+    }
+    pub fn get_merkle_proof(&self, license_hash: &BlsScalar) -> PoseidonBranch<DEPTH> {
+        let mut pos = 0;
+
+        for i in 0..(4 ^ DEPTH) {
+            let it = i.try_into().unwrap();
+
+            let leaf = self.tree.get(it);
+
+            match leaf {
+                Some(leaf) if leaf.license_hash == *license_hash => pos = it,
+                _ => (),
+            }
+        }
```

Comment:
this loop has a huge range, `4 ^ DEPTH`, and nothing seems to interrupt it or finish it early, are we sure that is what we want? We could interrupt the loop when pos is found, why is this not done?

Reply:
good question, I removed the break for testing and I forgot to readd it